### PR TITLE
Removing APMs from the legacy checkout experience

### LINKED ISSUE #3524: Remove APMs on the legacy checkout due deprecation
Stripe is deprecating APMs (Alternative Payment Methods) on the legacy checkout (using Sources API) by October 31st, 2024. Because of that, we need to remove them automatically from the settings page, and the checkout pages when UPE is disabled and the deadline has been reached.

### PR DESCRIPTION
**Pull request by wjrosa**
Fixes #3524

## Changes proposed in this Pull Request:

This PR introduces a series of changes related to the sunsetting of APMs on the legacy checkout experience (by Oct, 31st 2024):
- Slightly changes the APMs sunsetting promotional surface where it says "Stripe will end support..." to "Stripe ended support..."
- Blocks the enablement of APMs when the legacy checkout is enabled, and adds a notice above them about the deprecation
- Removes them from the checkout page

## Testing instructions

- Checkout and build this branch on your development environment (`tweak/remove-apms-legacy-checkout`)
- Hardcode the return value as `true` for both `are_apms_deprecated` (PHP) and `areAPMsDeprecated` (JS) functions
- Enable the legacy checkout experience

### Settings page

- As a merchant, confirm that you can see APMs disabled with a new pill on top of it explaining the deprecation on the settings page (`tweak/remove-apms-legacy-checkout`)
Screenshot 2024-10-18 at 17 13 43

### Checkout page

- As a merchant, create a shortcode checkout page
- As a shopper, add a product to your cart
- Go to the shortcode checkout page
- Confirm that you cannot see any APMs
- You can even enable all of them on the `develop` branch and go back to this one and see if they are still unavailable
Screenshot 2024-10-18 at 17 08 22

### Promotional surface

- As a merchant, confirm that you can see the updated copy:
Screenshot 2024-10-18 at 17 12 09

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [x] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

### CODE REVIEW

File: client/utils/index.js [hunk 1]
```diff
@@ -8,3 +8,13 @@ export const getAddToCartVariationParams = ( key ) => {
...
+
+/**
+ * Check if APMs are deprecated. Meaning it is past Oct 31st, 2024 and the legacy checkout is enabled.
+ *
+ * @return {boolean} Whether APMs are deprecated.
+ */
+export const areAPMsDeprecated = () =>
+	new Date() > new Date( '2024-10-31' ) &&
+	// eslint-disable-next-line camelcase
+	wc_stripe_settings_params.is_upe_checkout_enabled !== '1';
```
Comment (annemirasol) [inline client/utils/index.js]: What do you think about having a single place for this logic and just passing a `are_apms_deprecated` param from the backend in `wc_stripe_settings_params`?

**Comment by annemirasol** [comment]
> As a merchant, confirm that you can see APMs disabled with a new pill on top of it explaining the deprecation on the settings page (tweak/remove-apms-legacy-checkout)

- So I changed the cutoff date in the PHP and JS files (to `2024-01-31`), and turned on legacy checkout. This is what I see:

Screenshot 2024-10-21 at 3 31 20 PM

---

- With the original cutoff date (`2024-10-31`) and legacy checkout still enabled, I see this:
Screenshot 2024-10-21 at 3 33 26 PM

File: includes/class-wc-stripe-helper.php [hunk 2]
```diff
@@ -1601,4 +1605,13 @@ public static function get_klarna_preferred_locale( $store_locale, $billing_coun
...
 		return $target_locale;
 	}
+
+	/**
+	 * Checks if the APMs are deprecated. Stripe deprecated them by October 31, 2024 (for the legacy checkout).
+	 *
+	 * @return bool Whether the APMs are deprecated.
+	 */
+	public static function are_apms_deprecated() {
+		return ( new \DateTime() )->format( 'Y-m-d' ) > '2024-10-31' && ! WC_Stripe_Feature_Flags::is_upe_checkout_enabled();
```
Comment (diegocurbelo) [inline includes/class-wc-stripe-helper.php]: It should be Oct 29th (the same date mentioned in banner message)
```suggestion
		return ( new \DateTime() )->format( 'Y-m-d' ) > '2024-10-29' && ! WC_Stripe_Feature_Flags::is_upe_checkout_enabled();
```

File: client/utils/index.js [hunk 3]
```diff
@@ -8,3 +8,13 @@ export const getAddToCartVariationParams = ( key ) => {
...
 	return wcAddToCartVariationParams[ key ];
 };
+
+/**
+ * Check if APMs are deprecated. Meaning it is past Oct 31st, 2024 and the legacy checkout is enabled.
+ *
+ * @return {boolean} Whether APMs are deprecated.
+ */
+export const areAPMsDeprecated = () =>
+	new Date() > new Date( '2024-10-31' ) &&
```
Comment (diegocurbelo) [inline client/utils/index.js]: It should be Oct 29th (the same date mentioned in banner message)
```suggestion
	new Date() > new Date( '2024-10-29' ) &&
```

I like @annemirasol suggestion to rely on the server for this flag; that way, we don't need to worry about the customer's browser date being wrong 🤷🏼

**Review by diegocurbelo — CHANGES_REQUESTED** [review]

File: client/utils/index.js (hunk 1 above)
Comment (wjrosa) [reply client/utils/index.js]: Yeah, that's smarter! Done ✅

**Comment by wjrosa** [comment]
> > As a merchant, confirm that you can see APMs disabled with a new pill on top of it explaining the deprecation on the settings page (tweak/remove-apms-legacy-checkout)
> 
> * So I changed the cutoff date in the PHP and JS files (to `2024-01-31`), and turned on legacy checkout. This is what I see:
> 
> Screenshot 2024-10-21 at 3 31 20 PM
> * With the original cutoff date (`2024-10-31`) and legacy checkout still enabled, I see this:
> 
> Screenshot 2024-10-21 at 3 33 26 PM

Thanks for the review, Anne! I have fixed the issue with methods not being listed. And updated the JS code to use the backend flag instead of duplicating the logic there.

File: changelog.txt [hunk 4]
```diff
@@ -1,6 +1,7 @@
 *** Changelog ***
 
 = 8.9.0 - xxxx-xx-xx =
+* Tweak - Disables APMs when using the legacy checkout experience due Stripe deprecation by October 31st, 2024.
```
Comment (annemirasol) [inline changelog.txt]:
```suggestion
* Tweak - Disables APMs when using the legacy checkout experience due Stripe deprecation by October 29, 2024.
```

File: client/components/payment-method-deprecation-pill/index.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+const PaymentMethodDeprecationPill = () => {
+	return (
+		<>
+			<StyledPill>
+				{ __( 'Deprecated', 'woocommerce-gateway-stripe' ) }
+
+				<Popover
+					BaseComponent={ IconComponent }
+					content={ __(
+						'This payment method is deprecated on the legacy checkout as of Oct, 29th 2024.'
```
Comment (annemirasol) [inline client/components/payment-method-deprecation-pill/index.js]:
```suggestion
						'This payment method is deprecated on the legacy checkout as of Oct 29, 2024.'
```

File: readme.txt [hunk 6]
```diff
@@ -129,6 +129,7 @@ If you get stuck, you can ask for help in the Plugin Forum.
 == Changelog ==
 
 = 8.9.0 - xxxx-xx-xx =
+* Tweak - Disables APMs when using the legacy checkout experience due Stripe deprecation by October 31st, 2024.
```
Comment (annemirasol) [inline readme.txt]:
```suggestion
* Tweak - Disables APMs when using the legacy checkout experience due Stripe deprecation by October 29, 2024.
```

**Comment by annemirasol** [comment]
@wjrosa Looks great ✨ Thanks for working on this!

I spotted some typo regarding the new date and added suggested fixes.

Minor fix requested 🙏: When a notice box/modal is open, I expect that clicking anywhere outside the box should dismiss it. Other notices, like "Requires currency", behaves that way. "Deprecated" doesn't. Worth checking if it's an easy thing to add?

Screenshot 2024-10-22 at 10 19 59 AM

**Comment by wjrosa** [comment]
> @wjrosa Looks great ✨ Thanks for working on this!
> 
> I spotted some typo regarding the new date and added suggested fixes.
> 
> Minor fix requested 🙏: When a notice box/modal is open, I expect that clicking anywhere outside the box should dismiss it. Other notices, like "Requires currency", behaves that way. "Deprecated" doesn't. Worth checking if it's an easy thing to add?
> 
> Screenshot 2024-10-22 at 10 19 59 AM

Thanks for the suggestions! I've applied all of them. As for the popover bug, I fixed it by including a link with the Stripe page on it. The bug actually exists on the component itself (it works only when a link is present), but I thought it could be a nice solution.

File: client/components/payment-method-deprecation-pill/index.js [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+	return (
+		<StyledPill>
+			{ __( 'Deprecated', 'woocommerce-gateway-stripe' ) }
+			<Popover
+				BaseComponent={ IconComponent }
+				content={ interpolateComponents( {
+					mixedString:
+						/* translators: $1: a payment method name. %2: Currency(ies). */
+						__(
+							'This payment method is deprecated on the {{currencySettingsLink}}legacy checkout as of Oct, 29th 2024{{/currencySettingsLink}}.',
```
Comment (annemirasol) [inline client/components/payment-method-deprecation-pill/index.js]: Nit: Shouldn't the comma be after `29th`?

**Review by annemirasol — APPROVED** [review]
Thanks for fixing the modal dismiss issue! 🚢 

✅ APMs are deprecated if after the cutoff date and legacy checkout is enabled.
✅ APMs are not deprecated if after the cutoff date and legacy checkout is enabled.
✅ APMs are not deprecated if after the cutoff date and legacy checkout is not enabled.
✅ APMs are not deprecated if before the cutoff date.

File: client/components/payment-method-deprecation-pill/index.js (hunk 7 above)
Comment (wjrosa) [reply client/components/payment-method-deprecation-pill/index.js]: Oops, fixed it now!

**Pull request merged**